oauth.go

```diff
@@ -0,0 +1,172 @@
+package scalingo
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+
+	"golang.org/x/oauth2"
+	"gopkg.in/errgo.v1"
+)
+
+const defaultAuthUrl = "https://auth.scalingo.com"
+
+type TokenGenerator interface {
+	GetAccessToken() (string, error)
+}
+
+type OAuthTokenGenerator struct {
+	source       oauth2.TokenSource
+	AccessToken  string `json:"access_token"`
+	RefreshToken string `json:"refresh_token"`
+	TokenURL     string `json:"token_url"`
+	Token        string `json:"token"`
+}
+
+type OAuthApplication struct {
+	UID    string `json:"uid"`
+	Secret string `json:"secret"`
+}
+
+type LoginParams struct {
+	Identifier string `json:"identifier"`
+	Password   string `json:"password"`
+	OTP        string `json:"otp"`
+}
+
+type TokenResponse struct {
+	App   *OAuthApplication `json:"app"`
+	Token *Token            `json:"token"`
+}
+
+func (c *Client) IsOTPRequired(err error) bool {
+	rerr, ok := err.(*RequestFailedError)
+	if !ok {
+		return false
+	}
+
+	if rerr.Message == "OTP Required" {
+		return true
+	}
+	return false
+}
+
+func (c *Client) GetOAuthCredentials(params LoginParams) (*OAuthApplication, *Token, error) {
+	req := &APIRequest{
+		Client:   c,
+		NoAuth:   true,
+		Method:   "POST",
+		URL:      fmt.Sprintf("%s/v1/client/cli", c.AuthURL()),
+		Username: params.Identifier,
+		Password: params.Password,
+		OTP:      params.OTP,
+	}
+
+	resp, err := req.Do()
+	if err != nil {
+		if c.IsOTPRequired(err) {
+			return nil, nil, errgo.New("OTP Required")
```

Comment:
Isn't it better to use errors in variable like the MongoDB driver (https://godoc.org/gopkg.in/mgo.v2#pkg-variables) to be able to test it at line 50 (https://github.com/Scalingo/go-scalingo/pull/23/files#diff-b3c698f6274c18640399beb95db68658R50)